domain/removal/state/unit.go

```diff
@@ -65,62 +68,101 @@ AND    life_id = 0`, unitUUID)
 		return "", errors.Errorf("preparing unit life update: %w", err)
 	}
 
+	var mUUID string
+	if err := errors.Capture(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if err := tx.Query(ctx, updateUnitStmt, unitUUID).Run(); err != nil {
+			return errors.Errorf("advancing unit life: %w", err)
+		}
+
+		mUUID, err = st.markMachineAsDyingIfAllUnitsAreNotAlive(ctx, tx, uUUID)
+		if err != nil {
+			return errors.Errorf("marking last unit on machine as dying: %w", err)
+		}
+
+		return nil
+	})); err != nil {
+		return "", err
+	}
+
+	return mUUID, nil
+}
```

Comment:
I do not like this one bit. I don't think this should be included in `EnsureUnitNotAlive`, this is a pretty nasty side-effect for a method with such a simple, direct name.

Have you considered splitting this off into it's own state method?

Reply:
My preference would be:
- `EnsureUnitNotAlive` always returns the machineUUID for the machine it is assigned to (if there is one); and
- `markMachineAsDyingIfAllUnitsAreNotAlive` is made public, which is called in `RemoveUnit` service method

However, if you need to keep this to one transaction, I would be ok with re-naming this method to something longer and more cumbersome to capture it's complexity

Reply:
This has to be in one transaction. This is why the complexity rises. We must ensure that nothing else can attempt to use a unit/machine if they're in the dying state. There is also no way to transition back from dying/dead. Once you're there you've set off a one way sequence.